sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -357,11 +357,11 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
     val canUseV2 = lookupV2Provider().isDefined
 
     session.sessionState.sqlParser.parseMultipartIdentifier(tableName) match {
-      case CatalogObjectIdentifier(catalog, ident) if !isSessionCatalog(catalog) =>
+      case NonSessionCatalogAndIdentifier(catalog, ident) =>
```

Comment:
This looks much better (BTW, this is not a replacement from `CatalogObjectIdentifier` to `CatalogAndIdentifer`)

Reply:
I like this change. Could you update the PR description more precisely?

Reply:
Updated. Thanks!